### PR TITLE
Added support for a TUITableView footer view

### PR DESCRIPTION
**Pull request by jagreenwood**

### CODE REVIEW

File: lib/UIKit/TUITableView.h [hunk 1]
```diff
@@ -70,7 +70,8 @@
 	NSArray                     * _sectionInfo;
 	
 	TUIView                     * _pullDownView;
-	TUIView							        * _headerView;
+	TUIView						* _headerView;
+	TUIView						* _footerView;
```
Comment (avaidyam) [inline lib/UIKit/TUITableView.h]: No ivars please! Switch to private properties. (No need for `@synthesize` either....)

File: lib/UIKit/TUITableView.m [hunk 2]
```diff
@@ -907,6 +915,23 @@ - (void)_layoutCells:(BOOL)visibleCellsNeedRelayout
 			}
 		}
 	}
+	
+	if (_footerView) {
```
Comment (avaidyam) [inline lib/UIKit/TUITableView.m]: Indentation looks weird from here on down... might want to switch this to 4-spaced tabs.

File: lib/UIKit/TUITableView.m [hunk 3]
```diff
@@ -907,6 +915,23 @@ - (void)_layoutCells:(BOOL)visibleCellsNeedRelayout
...
+		CGRect footerViewRect = CGRectMake(0, 0, visible.size.width, _footerView.frame.size.height);
+		if(CGRectIntersectsRect(footerViewRect, visible)) {
+			_footerView.frame = footerViewRect;
+			[_footerView setNeedsLayout];
+			
+			if(_footerView.hidden) {
```
Comment (avaidyam) [inline lib/UIKit/TUITableView.m]: These two `if(_footerView.hidden)` statements can be reduced to a ternary `?:` operator statement. Simplicity is always best!

**Comment by avaidyam** [comment]
I've left a few comments on the PR so as to save @jspahrsummers and @jwilling some time, but this coincided perfectly with my requirements for a `TUITableFooterView`.... Thanks!

File: lib/UIKit/TUITableView.m [hunk 4]
```diff
@@ -907,6 +915,23 @@ - (void)_layoutCells:(BOOL)visibleCellsNeedRelayout
...
+				// show
+				_footerView.hidden = NO;
+			}
+		} else {
+			if(!_footerView.hidden) {
```
Comment (CodaFi) [inline lib/UIKit/TUITableView.m]: This `if()` statement is redundant, it is handled by the `else` block already.

File: lib/UIKit/TUITableView.h (hunk 1 above)
Comment (CodaFi) [inline lib/UIKit/TUITableView.h]: @galaxas0 Not necessarily a private property per se, seeing as he is declaring a backing iVar after all.  The public property should suffice.  It's also good to throw a smattering of documentation in along with whatever you add to the class.

File: lib/UIKit/TUITableView.m (hunk 3 above)
Comment (CodaFi) [inline lib/UIKit/TUITableView.m]: If simplicity is best, he could reduce it to `_footerView.hidden = !footerView.hidden;`

**Comment by jagreenwood** [comment]
For the most part I mimicked what was done for headerView for sake of consistency. I'll take a look at simplifying the setter. Cheers

**Comment by jwilling** [comment]
@jagreenwood Hey, thanks for your pull request. I agree consistency is important.

Since you're modifying the portion of the code that involves the headers and footers, it's a good time to be transitioning those over to properties with no implicit ivars and no synthesize. You've already done most of the work, so I think all that's needed at this point is simply to delete the ivars for those header and footer properties.

File: lib/UIKit/TUITableView.m (hunk 2 above)
Comment (jwilling) [inline lib/UIKit/TUITableView.m]: Along with the transition to no ivars in the header, Github's [coding conventions](https://github.com/github/objective-c-conventions) try to limit the use of ivars, unless explicitly needed (such as in -dealloc or -init). So in this conditional (as well as the ones below), could you use the property instead of the backing ivar?

I realize that not all the rest of the code does this, but this is just a practice put in place on new code. :+1:

Comment (jwilling) [inline lib/UIKit/TUITableView.m]: @galaxas0 The indentation is fine. He's using tabs.

Comment (avaidyam) [inline lib/UIKit/TUITableView.m]: @jwilling Alright, I think I'm unable to see that on github.com then. Weird...?

File: lib/UIKit/TUITableView.h (hunk 1 above)
Comment (avaidyam) [inline lib/UIKit/TUITableView.h]: @CodaFi Ah! Didn't catch that. Thanks :)

File: lib/UIKit/TUITableView.m (hunk 3 above)
Comment (avaidyam) [inline lib/UIKit/TUITableView.m]: Or, since a bool is an `unsigned char`, you could, use this: `_footerView.hidden *= -1;`

Comment (jwilling) [inline lib/UIKit/TUITableView.m]: @galaxas0 Nope, we're not going to do that. :see_no_evil:

Comment (jagreenwood) [inline lib/UIKit/TUITableView.m]: Since _layoutCells: is being called constantly while the table is being scrolled, I'm seeing that hidden has to be set explicitly or the footer view blinks.

File: lib/UIKit/TUITableView.m [hunk 5]
```diff
@@ -907,6 +915,17 @@ - (void)_layoutCells:(BOOL)visibleCellsNeedRelayout
 			}
 		}
 	}
+	
+	if (self.footerView) {
+		CGRect footerViewRect = CGRectMake(0, 0, visible.size.width, self.footerView.frame.size.height);
+		if(CGRectIntersectsRect(footerViewRect, visible)) {
+			self.footerView.frame = footerViewRect;
+			[self.footerView setNeedsLayout];
+			self.footerView.hidden = NO;
```
Comment (jwilling) [inline lib/UIKit/TUITableView.m:917]: Thanks for updating your pull request. I should have spoken up before, but in your previous commit you were simply being consistent with the header view code that was immediately before your code. Sorry to make you have to create another commit, but can you either revert this section back to the previous checks to match the style above, or alternatively update the header view code to match this?

Comment (jagreenwood) [inline lib/UIKit/TUITableView.m:917]: Sure, no prob!

**Comment by jwilling** [comment]
Thanks for your work on this @jagreenwood. Just a couple things need to be dealt with. The first is the note I added above.

Secondly, with all new additions to the user interface it is our goal to have the demo updated to reflect those changes in some way. Is it possible you could add onto the demo to demonstrate this functionality? Thanks.

**Comment by jagreenwood** [comment]
@jwilling Sure thing, I'll fix these up in 2 separate commits.

**Comment by jwilling** [comment]
Thanks, @jagreenwood!
